Conformal inference documentation (#23)
tidymodels/tidymodels.org: learn/models/conformal-regression/index.qmd

---
title: "Conformal inference for regression models"
categories:
  - probably
  - parsnip
  - tune
  - regression
type: learn-subsection
weight: 5
description: | 
  The probably package has functions to create prediction intervals for regression models.
toc: true
toc-depth: 2
include-after-body: ../../../resources.html
---

```{r}
#| label: "setup"
#| include: false
#| message: false
#| warning: false
source(here::here("common.R"))
```

```{r}
#| label: startup
#| include: false
#| cache: false
library(tidymodels)
library(probably)
library(doMC)

# -------------------------------------------------------------------------

tidymodels_prefer()
theme_set(theme_bw())
options(pillar.advice = FALSE, pillar.min_title_chars = Inf)
registerDoMC(cores = parallel::detectCores())

# ------------------------------------------------------------------------------

pkgs <- c("tidymodels", "probably", "nnet")
```

`r article_req_pkgs(pkgs)` The probably package should be version 1.0.1 or greater.

> What is [conformal inference](https://en.wikipedia.org/wiki/Conformal_prediction)? It is a collection of statistical methods that are mostly used to construct prediction intervals (or prediction sets) for any type of regression or classification model. The basic idea revolves around some theory on how to construct probability statements for a population of data points. 

Reviewer: That last sentence about the basic idea is very broad/vague

Author: I rewrote it

Since this article is focused on regression problems, we'll motivate the methodology in the context of a numeric outcome. For example, suppose we have collected a set of 1,000 standard normal data points. 

```{r}
#| label: normal-dist
#| out-width: 70%
#| fig-width: 6
#| fig-height: 4

library(tidymodels)
tidymodels_prefer()

set.seed(1)
reference_data <- tibble(data = rnorm(1000))

reference_data %>% 
  ggplot(aes(x = data)) +
  geom_line(stat = "density")
```

If we had a new observation that we thought might be from the same distribution, how would we say (probabilistically) whether we believe that it belongs to the original distribution? 

If we thought that 1,000 were a sufficient sample size, we might compute some quantiles of these data to define "the mainstream of the data." Let's use the 5th and 95th quantiles to set boundaries that define what we would expect to see most of the time:   

```{r}
#| label: normal-quant
#| out-width: 70%
#| fig-width: 6
#| fig-height: 4
quants <- quantile(reference_data$data, probs = c(0.05, 0.95))

reference_data %>% 
  ggplot(aes(x = data)) +
  geom_line(stat = "density") + 
  geom_vline(xintercept = quants, col = "red", lty = 2)
```

> If we were to get a new sample beyond these data, we would be able to say that we are about 90% sure that the new sample does not conform to the original distribution. This works under the assumption that the data are exchangeable. 

Reviewer: ```suggestion
If we were to get a new sample beyond these boundaries, we would be able to say that we are about 90% sure that the new sample does not conform to the original distribution. This works under the assumption that the data are exchangeable. 
```

Reviewer: I'm getting stuck on this sentence.

If we are talking confidence intervals, I think that sentence is skirting between "the interval covers the parameter" and "the parameter lies in the interval". But since we are talking about prediction intervals, does this still apply? 

What I'm trying to say: is this sentence statistically correct? I can't judge that immediately and my training on how to talk about p-values is kicking in - but it might be misplaced here.

Author: I rewrote this (emphasis added here to show the differences): 

> If we compute a prediction on a new unknown sample, we could center this reference distribution around its predicted value. For some _significance_ level, we now know the range of _sample_ values that "conform" to the _variance seen in the_ reference distribution. That range can define our prediction interval.

Author: The use of "predicted values" instead of "sample values" was incorrect.

We can apply this relatively simple idea to model predictions. Suppose we have a model created on a numeric outcome. If we make predictions on a data set we can compute the model residuals and create a sort of reference error distribution. If we compute a prediction on a new unknown sample, we could center this reference distribution around its predicted value. For some confidence level, we now know the range of predicted values that "conform" to the reference distribution. That range can define our prediction interval. 

> There are a variety of ways to apply this concept (which is unsurprisingly more complex than that the above description). The probably package has implemented a few of these.

Reviewer: ```suggestion
There are a variety of ways to apply this concept (which is unsurprisingly more complex than the above description). The probably package has implemented a few of these.
```

```{r}
#| label: example-data
#| out-width: 70%
#| fig-width: 6
#| fig-height: 4
make_data <- function(n, std_dev = 1 / 5) {
  tibble(x = runif(n, min = -1)) %>%
    mutate(
      y = (x^3) + 2 * exp(-6 * (x - 0.3)^2),
      y = y + rnorm(n, sd = std_dev)
    )
}

n <- 1000
set.seed(8383)
train_data <- make_data(n)

train_data %>% 
  ggplot(aes(x, y)) + 
  geom_point(alpha = 1 / 10)
```

We'll use these data as a training set and fit a model: 

```{r}
#| label: model-fit
#| out-width: 70%
#| fig-width: 6
#| fig-height: 4

set.seed(484)
nnet_wflow <- 
  workflow(y ~ x, mlp(hidden_units = 4) %>% set_mode("regression"))

nnet_fit <- nnet_wflow %>% fit(train_data)
nnet_pred <- augment(nnet_fit, train_data)

train_data %>% 
  ggplot(aes(x)) + 
  geom_point(aes(y = y), alpha = 1 / 10) +
  geom_line(data = nnet_pred, aes(y = .pred),
            linewidth = 1, col = "blue")
```

Let's examine three methods^[A fourth method is implemented for "full conformal prediction." It was developed mainly as a proof of concept. While it is effective, it is _very_ computationally expensive.] to produce prediction intervals: 

## Split Conformal Inference

> The most straightforward approach is reserving some data for estimating the residual distribution. We know that simply re-predicting the training set is **a bad idea**; the residuals would be smaller than they should be since the same data are used to create and evaluate the model. 

Reviewer: It'd be nice if this section makes it more clear why this method is called split conformal inference. I'm guessing it's because we split the data here, where you call it reserving data?

> Let's simulate another data set containing 250 samples and call that the "calibration set". These data can be predicted and the corresponding residuals can be used to define what conforms to model. We'll also create a large test set to see if we've done a good job. 

Reviewer: ```suggestion
Let's simulate another data set containing 250 samples and call that the "calibration set". These data can be predicted and the corresponding residuals can be used to define what conforms to the model. We'll also create a large test set to see if we've done a good job. 
```

split_int
```

To get predictions on new data, we use the standard `predict()` method on this object: 

```{r}
#| label: split-pred

# Produce 90% prediction intervals
test_split_res <- 
  predict(split_int, test_data, level = 0.90) %>% 
  bind_cols(test_data)

test_split_res %>% slice(1:5)
```

The results: 

```{r}
#| label: split-pred-plit
#| out-width: 70%
#| echo: false
#| fig-width: 6
#| fig-height: 4

test_split_res %>% 
  ggplot(aes(x)) +
  geom_point(aes(y = y), alpha = 1 / 20) +
  geom_line(
    data = nnet_pred, aes(y = .pred),
    linewidth = 1, col = "blue"
  ) +
  geom_ribbon(aes(ymin = .pred_lower, ymax = .pred_upper), 
              col = "#D95F02", linewidth = 3 / 4, fill = NA)

cal_resid_width <- 
  quantile(test_split_res$y - test_split_res$.pred, probs = c(0.05, 0.95)) %>% 
  diff() %>% 
  signif(3)
```

Since we know the outcome values, we can compute the coverage for this particular data set. Since we created a 90% prediction interval, about 90% of the outcomes should be within our bounds. Let's create a function and apply it to these data:

```{r}
#| label: split-coverage
coverage <- function(x) {
  x %>% 

>     mutate(in_bound = .pred_lower < y & .pred_upper > y) %>% 

Reviewer: Should this be inclusive of the boundaries?

    summarise(coverage = mean(in_bound) * 100)
}
coverage(test_split_res)
```

## Using Resampling Results

> In a way, the calibration set serves a similar role as a traditional validation set. Since we can just re-predict our training set, we need to evaluate our model on a separate collection of labeled data. 

Reviewer: non-native speaker question: do you serve a role or is that a mash-up of serving a purpose and having a role?

> In a way, the calibration set serves a similar role as a traditional validation set. Since we can just re-predict our training set, we need to evaluate our model on a separate collection of labeled data. 

Reviewer: ```suggestion
In a way, the calibration set serves a similar role as a traditional validation set. Since we can't just re-predict our training set, we need to evaluate our model on a separate collection of labeled data. 
```

> However, we also know that resampling methods can serve the same purpose. During resampling, we can compute a set of predictions that were not used to fit the model. For example, in 10-fold cross-validation, there are 10 collections of held-out predictions that do not suffer from the severe bias that occurs when we simple re-predict the training set. 

Reviewer: The "however" feels a bit out of place to me

The CV+ estimator ([Barber _et al._ (2021)](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=%22Predictive+inference+with+the+jackknife%2B%22&btnG=)) can be used to assemble a reference distribution of residuals. If we use `fit_resamples()` or one of the `tune_*()` functions, we can gather those residuals and use them to create prediction intervals. We'll have to make sure that we save the out-of-sample predictions as well as the resasmpled models.  

The `control_*()` functions can be used for this purpose. We can set `save_pred = TRUE` to save the predicted values. For the resampled models, we can use the tools to [extract the models](https://tidymodels.github.io/tidymodels_dot_org/learn/index.html#category=extracting%20results) via the `extract` argument. We'll use the `I()` function to return the fitted workflows from each resample: 

```{r}
#| label: control
ctrl <- control_resamples(save_pred = TRUE, extract = I)
```

Let's use 10-fold cross-validation to resample the neural network: 

```{r}
#| label: cv

set.seed(493)
folds <- vfold_cv(train_data)

nnet_rs <- 
  nnet_wflow %>% 
  fit_resamples(folds, control = ctrl)

collect_metrics(nnet_rs)
```

> The model has an estimated root mean squared error of `r collect_metrics(nnet_rs) %>% dplyr::filter(.metric == "rmse") %>% pluck("mean") %>% signif(3)` and an estimated R<sup>2</sup> of `r collect_metrics(nnet_rs) %>% dplyr::filter(.metric == "rsq") %>% pluck("mean") %>% signif(3)`.

Reviewer: Is this to get a general sense for how the model is doing? If so, it might be helpful to say that, so that people don't assume that this is necessary for getting prediction intervals.

nnet_variable_rs <- 
  nnet_wflow %>% 
  fit_resamples(folds_variable, control = ctrl)

cv_variable_int <- int_conformal_cv(nnet_variable_rs)

test_cv_variable_res <- 
  predict(cv_variable_int, test_variable_data, level = 0.90) %>% 
  bind_cols(test_variable_data)

test_cv_variable_res %>% 
  ggplot(aes(x)) +
  geom_point(aes(y = y), alpha = 1 / 20) +
  geom_line(
    data = nnet_variable_pred, aes(y = .pred),
    linewidth = 1, col = "blue"
  ) +
  geom_ribbon(aes(ymin = .pred_lower, ymax = .pred_upper), 
              col = "#D95F02", linewidth = 1, fill = NA)
```

While the coverage is good (`r round(coverage(test_cv_variable_res)$coverage, 1)`%), there are obvious areas where it is either too wide or too narrow. 

Conformalized quantile regression ([Romano _et al_ (2019)](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=%22Conformalized+quantile+regression%22)) is a method to produce intervals that can properly scale the intervals based on what was observed in the training data. It fits an initial [quantile regression](https://en.wikipedia.org/wiki/Quantile_regression) model to do so and also required a split data set, such as our calibration data. 

> The original work used basic quantile regression models. For tidymodels, we'll use [quantile regression forests](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=%22Quantile+Regression+Forests%22). The reason being that some models require nonlinear terms, interactions, and other features to model the data adequately. Users would have to have a parallel modeling process for that model. Random forests for quantile regression are very adaptable, low maintenance, and are robust to tuning parameter specification. There is a minor downside, as seen below. 

Reviewer: I really like this. It might be worth it to be a little bit more explicit about that that minor downside is, i.e. refer back to this sentence when you talk about the bumpiness of the intervals?

The function for this, `int_conformal_quantile()`, has a slightly different interface than the previous functions: 

* It requires both the training and calibration sets.
* The confidence level must be set in advance. 

> We'll also pass an argument to the `quantregForest()` function that will use 2,000 trees to make the forest: 

Reviewer: ```suggestion
We'll also pass an argument to the `quantregForest()` function that we'll use 2,000 trees to make the forest: 
```

```{r}
#| label: quant-pred

quant_int <-
  int_conformal_quantile(
    nnet_variable_fit, 
    train_data = train_variable_data,
    cal_data = cal_variable_data, 
    level = 0.90,
    ntree = 2000)
quant_int

test_quant_res <- 
  predict(quant_int, test_variable_data) %>% 
  bind_cols(test_variable_data)

test_quant_res %>% slice(1:5)
```

The results: 

```{r}
#| label: quant-pred-split
#| echo: false
#| out-width: 70%
#| fig-width: 6
#| fig-height: 4

test_quant_res %>% 
  ggplot(aes(x)) +
  geom_point(aes(y = y), alpha = 1 / 20) +
  geom_line(
    data = nnet_variable_pred, aes(y = .pred),
    linewidth = 1, col = "blue"
  ) +
  geom_ribbon(aes(ymin = .pred_lower, ymax = .pred_upper), 
              col = "#D95F02", linewidth = 3 / 4, fill = NA)
```

The bumpiness of the bound is the consequence of using a tree-based model for the quantiles. Despite that, the intervals do adapt their widths to suit the data. The coverage is also close to the target: `r round(coverage(test_quant_res)$coverage, 1)`%. 

> ## Coverage

Reviewer: This sounds a lot more casual than other parts of the text.

Speaking of coverage, there's a [GitHub repo](https://github.com/topepo/conformal_sim) where we did simulations for these methods to determine their average coverage. The repository's readme file has a summary. 

## Looking Forward

> This is a growing research area; we will do our best to implement important new methods as they are created. 

Reviewer: I think we usually don't put such forward-looking statements in the tidymodels.org articles?  That seems much more suited for a blog post, maybe with a link to the GH issue to suggest particularly great new methods.

> If you are interested and would like to learn more, try these resources: 

Reviewer: While it's great to have places to learn more, this reminds me of "Never be the keeper of the list".

#| out-width: 70%
#| fig-width: 6
#| fig-height: 4

library(tidymodels)
tidymodels_prefer()

set.seed(1)
reference_data <- tibble(data = rnorm(1000))

reference_data %>% 
  ggplot(aes(x = data)) +
  geom_line(stat = "density")
```

If we had a new observation that we thought might be from the same distribution, how would we say (probabilistically) whether we believe that it belongs to the original distribution? 

If we thought that 1,000 were a sufficient sample size, we might compute some quantiles of these data to define "the mainstream of the data." Let's use the 5th and 95th quantiles to set boundaries that define what we would expect to see most of the time:   

```{r}
#| label: normal-quant
#| out-width: 70%
#| fig-width: 6
#| fig-height: 4
quants <- quantile(reference_data$data, probs = c(0.05, 0.95))

reference_data %>% 
  ggplot(aes(x = data)) +
  geom_line(stat = "density") + 
  geom_vline(xintercept = quants, col = "red", lty = 2)
```

If we were to get a new sample beyond these data, we would be able to say that we are about 90% sure that the new sample does not conform to the original distribution. This works under the assumption that the data are exchangeable. 

> We can apply this relatively simple idea to model predictions. Suppose we have a model created on a numeric outcome. If we make predictions on a data set we can compute the model residuals and create a sort of reference error distribution. If we compute a prediction on a new unknown sample, we could use center this reference distribution around its predicted value. For some confidence level, we now know the range of predicted values that "conform" to the reference distribution. That range can define our prediction interval. 

Reviewer 2: ```suggestion
We can apply this relatively simple idea to model predictions. Suppose we have a model created on a numeric outcome. If we make predictions on a data set we can compute the model residuals and create a sort of reference error distribution. If we compute a prediction on a new unknown sample, we could center this reference distribution around its predicted value. For some confidence level, we now know the range of predicted values that "conform" to the reference distribution. That range can define our prediction interval. 
```

> There are a variety of ways to apply this concept (which is unsurprisingly more complex than that the above description). The probably package has implemented a few of them for regression models (classification methods are coming in the future). 

Reviewer 2: ```suggestion
There are a variety of ways to apply this concept (which is unsurprisingly more complex than that the above description). The probably package has implemented a few of these.
```

The sentence works for a blogpost/announcement. But I think it should be written more generally so we don't have to go back and update this page once we add classification methods

Let's make a simple example to illustrate the results. We'll simulate a data set with a single predictor along with some unknown samples: 

```{r}
#| label: example-data
#| out-width: 70%
#| fig-width: 6
#| fig-height: 4
make_data <- function(n, std_dev = 1 / 5) {
  tibble(x = runif(n, min = -1)) %>%
    mutate(
      y = (x^3) + 2 * exp(-6 * (x - 0.3)^2),
      y = y + rnorm(n, sd = std_dev)
    )
}

n <- 1000
set.seed(8383)
train_data <- make_data(n)

train_data %>% 

>   ggplot(aes(x, y )) + 

Reviewer 2: ```suggestion
  ggplot(aes(x, y)) + 
```

## Split Conformal Inference

The most straightforward approach is reserving some data for estimating the residual distribution. We know that simply re-predicting the training set is **a bad idea**; the residuals would be smaller than they should be since the same data are used to create and evaluate the model. 

Let's simulate another data set containing 250 samples and call that the "calibration set". These data can be predicted and the corresponding residuals can be used to define what conforms to model. We'll also create a large test set to see if we've done a good job. 

```{r}
#| label: more-data

set.seed(7292)
cal_data  <- make_data(250)
test_data <- make_data(10000)
```

The probably package has a set of functions with the prefix `int_conformal` that can be used to create prediction intervals. One is: 

```{r}
#| label: split-obj

split_int <- int_conformal_split(nnet_fit, cal_data)
split_int
```

To get predictions on new data, we use the standard `predict()` method on this object: 

```{r}
#| label: split-pred

# Produce 90% prediction intervals
test_split_res <- 

>   predict(split_int, test_data, level = 0.90) %>% 

Reviewer 2: This smells like it should have a `augment()` method

Author: We don't have one implemented yet but that's a great idea.

We can create another object for computing the intervals:

```{r}
#| label: int-cv

cv_int <- int_conformal_cv(nnet_rs)
cv_int

# Produce 90% prediction intervals
test_cv_res <- 
  predict(cv_int, test_data, level = 0.90) %>% 
  bind_cols(test_data)

test_cv_res %>% slice(1:5)
```

The results for this method are:

```{r}
#| label: cv-pred
#| out-width: 70%
#| echo: false
#| fig-width: 6
#| fig-height: 4

test_cv_res %>% 
  ggplot(aes(x)) +
  geom_point(aes(y = y), alpha = 1 / 20) +
  geom_line(
    data = nnet_pred, aes(y = .pred),
    linewidth = 1, col = "blue"
  ) +
  geom_ribbon(aes(ymin = .pred_lower, ymax = .pred_upper), 
              col = "#D95F02", linewidth = 3 / 4, fill = NA)
```

At each point, the interval length is `r round(mean(test_cv_res$.pred_upper - test_cv_res$.pred_lower), 2)`; the previous split conformal interval width was `r round(mean(test_split_res$.pred_upper - test_split_res$.pred_lower), 2)`. This is due to the training set being larger than the calibration set. 

> Note that the coverage is a little better too: 

Reviewer 2: Could you clarify what you mean by "better" here. I read it as "closer to 90%", is that correct?

```{r}
#| label: cv-coverage
coverage(test_cv_res)
```

## Adaptive width intervals

The simulated data that we've been using has a constant error variance. That has worked for the two methods shown since their intervals are always the same width. Real data does not always have constant variation. 

To demonstrate, we can take the previous simulation system and induce a variance that is dynamic over the predictor range: 

```{r}
#| label: variable-plot
#| out-width: 70%
#| fig-width: 6
#| fig-height: 4

make_variable_data <- function(n, std_dev = 1 / 5) {
  tibble(x = runif(n, min = -1)) %>%
    mutate(
      y = (x^3) + 2 * exp(-6 * (x - 0.3)^2),
      y = y + rnorm(n, sd = std_dev * abs(x))
    )
}

make_variable_data(1000) %>% 
  ggplot(aes(x, y)) + 
  geom_point(alpha = 1 / 5)
```

Let's create new data sets and re-fit our model: 

```{r}
#| label: variable-data
set.seed(7292)
train_variable_data  <- make_variable_data(n)
cal_variable_data  <- make_variable_data(250)

> test_variable_data <- make_variable_data(10000)

Reviewer 2: nit-pick: either align all the `<-` or none of them 😄

nnet_variable_fit <- nnet_wflow %>% fit(train_variable_data)
nnet_variable_pred <- augment(nnet_variable_fit, train_variable_data)
```

We can recreate the CV+ interval for this new version of the data:

```{r}
#| label: cv-redux
#| out-width: 70%
#| echo: false
#| fig-width: 6
#| fig-height: 4 
set.seed(493)
folds_variable <- vfold_cv(train_variable_data)

nnet_variable_rs <- 
  nnet_wflow %>% 
  fit_resamples(folds_variable, control = ctrl)

cv_variable_int <- int_conformal_cv(nnet_variable_rs)

test_cv_variable_res <- 
  predict(cv_variable_int, test_variable_data, level = 0.90) %>% 
  bind_cols(test_variable_data)

test_cv_variable_res %>% 
  ggplot(aes(x)) +
  geom_point(aes(y = y), alpha = 1 / 20) +
  geom_line(
    data = nnet_variable_pred, aes(y = .pred),
    linewidth = 1, col = "blue"
  ) +
  geom_ribbon(aes(ymin = .pred_lower, ymax = .pred_upper), 
              col = "#D95F02", linewidth = 1, fill = NA)
```

> While the coverage is good (`r round(coverage(test_cv_variable_res)$coverage, 1)`%), there are obvious areas where it is either too wide or too narrow. 

Reviewer 2: Is there a reason why you stopped running `coverage(test_cv_variable_res)` for people to see?

Author: Just to be more concise.

Conformalized quantile regression ([Romano _et al_ (2019)](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=%22Conformalized+quantile+regression%22)) is a method to produce intervals that can properly scale the intervals based on what was observed in the training data. It fits an initial [quantile regression](https://en.wikipedia.org/wiki/Quantile_regression) model to do so and also required a split data set, such as our calibration data. 

> The original work used basic quantile regression models. For tidymodels, we'll use [quantile regression forests](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=%22Quantile+Regression+Forests%22). The reason being that some models require nonlinear terms, interactions, and other features to model the data adequately. Users would have to have a parallel modeling process for that model. Random forests for quantile regression are very adaptable, are low maintenance, and are robust to tuning parameter specification. There is a minor downside, as seen below. 

Reviewer 2: I don't know if this is a good change, but it sounds better to me

```suggestion
The original work used basic quantile regression models. For tidymodels, we'll use [quantile regression forests](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=%22Quantile+Regression+Forests%22). The reason being that some models require nonlinear terms, interactions, and other features to model the data adequately. Users would have to have a parallel modeling process for that model. Random forests for quantile regression are very adaptable, low maintenance, and are robust to tuning parameter specification. There is a minor downside, as seen below. 
```

    ntree = 2000)
quant_int

test_quant_res <- 
  predict(quant_int, test_variable_data) %>% 
  bind_cols(test_variable_data)

test_quant_res %>% slice(1:5)
```

The results: 

```{r}
#| label: quant-pred-split
#| echo: false
#| out-width: 70%
#| fig-width: 6
#| fig-height: 4

test_quant_res %>% 
  ggplot(aes(x)) +
  geom_point(aes(y = y), alpha = 1 / 20) +
  geom_line(
    data = nnet_variable_pred, aes(y = .pred),
    linewidth = 1, col = "blue"
  ) +
  geom_ribbon(aes(ymin = .pred_lower, ymax = .pred_upper), 
              col = "#D95F02", linewidth = 3 / 4, fill = NA)
```

The bumpiness of the bound is the consequence of using a tree-based model for the quantiles. Despite that, the intervals do adapt their widths to suit the data. The coverage is also close to the target: `r round(coverage(test_quant_res)$coverage, 1)`%. 

## Coverage

Speaking of coverage, there's a [GitHub repo](https://github.com/topepo/conformal_sim) where we did simulations for these methods to determine their average coverage. The repository's readme file has a summary. 

## Looking Forward

This is a growing research area; we will do our best to implement important new methods as they are created. 

> Also, there are a multitude of methods for classification models. Those are next on our list. 

Reviewer 2: Same as before, I don't think this line to good to have on a long term basis
